src/pool-bin/lens/BinQuoter.sol

```diff
@@ -9,222 +9,159 @@ import {BalanceDelta} from "pancake-v4-core/src/types/BalanceDelta.sol";
+            (PoolKey memory poolKey, bool oneForZero) = pathKey.getPoolAndSwapDirection(outputCurrency);
+
+            swapDelta = _swap(poolKey, !oneForZero, amountOut.safeInt128(), pathKey.hookData);
+
+            amountOut = oneForZero ? uint128(-swapDelta.amount1()) : uint128(-swapDelta.amount0());
+            outputCurrency = pathKey.intermediateCurrency;
         }
+        // amountOut after the loop exits actually holds the amountIn of the trade
+        amountOut.revertQuote();
     }
 
     /// @dev quote an ExactOutput swap on a pool, then revert with the result
-    function _quoteExactOutputSingle(QuoteExactSingleParams memory params)
-        public
-        override
-        selfOnly
-        returns (bytes memory)
-    {
-        amountOutCached = params.exactAmount;
-
-        (BalanceDelta deltas, uint24 activeIdAfter) =
+    function _quoteExactOutputSingle(QuoteExactSingleParams calldata params) external selfOnly returns (bytes memory) {
+        BalanceDelta swapDelta =
             _swap(params.poolKey, params.zeroForOne, params.exactAmount.safeInt128(), params.hookData);
 
-        if (amountOutCached != 0) delete amountOutCached;
-        int128[] memory deltaAmounts = new int128[](2);
-
-        deltaAmounts[0] = deltas.amount0();
-        deltaAmounts[1] = deltas.amount1();
-
-        bytes memory result = abi.encode(deltaAmounts, activeIdAfter);
-        assembly ("memory-safe") {
-            revert(add(0x20, result), mload(result))
-        }
+        // the input delta of a swap is negative so we must flip it
+        uint256 amountIn = params.zeroForOne ? uint128(-swapDelta.amount0()) : uint128(-swapDelta.amount1());
+        amountIn.revertQuote();
     }
 
-    /// @dev Execute a swap and return the amounts delta, as well as relevant pool state
+    /// @dev Execute a swap and return the balance delta
     /// @notice if amountSpecified < 0, the swap is exactInput, otherwise exactOutput
     function _swap(PoolKey memory poolKey, bool zeroForOne, int128 amountSpecified, bytes memory hookData)
         private
-        returns (BalanceDelta deltas, uint24 activeIdAfter)
+        returns (BalanceDelta deltas)
     {
         deltas = poolManager.swap(poolKey, zeroForOne, amountSpecified, hookData);
 
-        // only exactOut case
-        if (amountOutCached != 0 && amountOutCached != uint128(zeroForOne ? deltas.amount1() : deltas.amount0())) {
-            revert InsufficientAmountOut();
-        }
-
-        (activeIdAfter,,) = poolManager.getSlot0(poolKey.toId());
+        // TODO: confirm we don't need this check in the bin quoter since BinPool will revert if there is not enough liquidity
+        // Check that the pool was not illiquid.
+        // int128 amountSpecifiedActual = (zeroForOne == (amountSpecified < 0)) ? deltas.amount0() : deltas.amount1();
+        // if (amountSpecifiedActual != amountSpecified) {
+        //     revert NotEnoughLiquidity(poolKey.toId());
+        // }
```

Comment:
Will remove comments once confirmed

Reply:
is this no longer applicable? https://github.com/pancakeswap/pancake-v4-periphery/blob/main/src/pool-bin/lens/BinQuoter.sol#L223-L226

Reply:
1. is this no longer applicable? https://github.com/pancakeswap/pancake-v4-periphery/blob/main/src/pool-bin/lens/BinQuoter.sol#L223-L226

Yes, i think even in previous version this will never be triggered.. 

2. or is it because BinPool will fire BinPool__OutOfLiquidity, so we no longer need to check here?

Yes, can imagine if amountOut is less than expected then it must because of the illiquidity of the pool. Then it should revert with BinPool__OutOfLiquidity before the check

Reply:
Oh, it just hit me hook might have influence on this. Let add it anyway, wdyt ? 😂

Reply:
agree